packages/block-editor/src/hooks/block-bindings.js

```diff
@@ -2,42 +2,164 @@
  * WordPress dependencies
  */
 import { __ } from '@wordpress/i18n';
-import { store as blocksStore } from '@wordpress/blocks';
+import { privateApis as blocksPrivateApis } from '@wordpress/blocks';
 import {
-	BaseControl,
-	PanelBody,
-	__experimentalHStack as HStack,
+	Button,
+	ButtonGroup,
 	__experimentalItemGroup as ItemGroup,
 	__experimentalItem as Item,
+	__experimentalText as Text,
+	__experimentalToolsPanel as ToolsPanel,
+	__experimentalToolsPanelItem as ToolsPanelItem,
+	privateApis as componentsPrivateApis,
 } from '@wordpress/components';
-import { useSelect } from '@wordpress/data';
+import { useSelect, useDispatch, useRegistry } from '@wordpress/data';
+import { useContext } from '@wordpress/element';
+import { useViewportMatch } from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
-import { canBindAttribute } from '../hooks/use-bindings-attributes';
+import {
+	canBindAttribute,
+	getBindableAttributes,
+} from '../hooks/use-bindings-attributes';
+import { store as blockEditorStore } from '../store';
 import { unlock } from '../lock-unlock';
 import InspectorControls from '../components/inspector-controls';
+import BlockContext from '../components/block-context';
+
+const {
+	DropdownMenuV2: DropdownMenu,
+	DropdownMenuGroupV2: DropdownMenuGroup,
+	DropdownMenuCheckboxItemV2: DropdownMenuCheckboxItem,
+	DropdownMenuItemLabelV2: DropdownMenuItemLabel,
+	DropdownMenuItemHelpTextV2: DropdownMenuItemHelpText,
+	DropdownMenuSeparatorV2: DropdownMenuSeparator,
+} = unlock( componentsPrivateApis );
+
+const useToolsPanelDropdownMenuProps = () => {
+	const isMobile = useViewportMatch( 'medium', '<' );
+	return ! isMobile
+		? {
+				popoverProps: {
+					placement: 'left-start',
+					// For non-mobile, inner sidebar width (248px) - button width (24px) - border (1px) + padding (16px) + spacing (20px)
+					offset: 259,
+				},
+		  }
+		: {};
+};
+
+function BlockBindingsPanelDropdown( {
+	fieldsList,
+	addConnection,
+	removeConnection,
+	attribute,
+	binding,
+} ) {
+	const currentKey = binding?.args?.key;
+	return (
+		<>
+			{ Object.entries( fieldsList ).map( ( [ label, fields ] ) => (
+				<>
+					<DropdownMenuGroup key={ label }>
+						{ Object.keys( fieldsList ).length > 1 && (
+							<Text
+								className="block-editor-bindings__source-label"
+								upperCase
+								variant="muted"
+							>
+								{ label }
+							</Text>
+						) }
+						{ Object.entries( fields ).map( ( [ key, value ] ) => (
+							<DropdownMenuCheckboxItem
+								key={ key }
+								onClick={ () =>
+									addConnection( key, attribute )
+								}
+								name={ attribute + '-binding' }
+								value={ key }
+								checked={ key === currentKey }
+							>
+								<DropdownMenuItemLabel numberOfLines={ 1 }>
```

Comment:
Even if it currently works, please do not use the `numberOfLines` prop. It's currently not part of `DropdownMenuItemLabel` and `DropdownMenuItemHelpText` public APIs and it could lead to unexpected behaviour / regressions in case we update the component in the future.

You won't get TypeScript autocompletion at the moment because of `lock/unlock` not supporting it, but you can have a look at the public API of `DropdownMenuV2` and its subcomponents in the [Storybook docs](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs)

Reply:
> Even if it currently works, please do not use the numberOfLines prop.

Is it possible to truncate on the first line? I couldn't find another way of doing that. Is it something you consider adding in the future?

Reply:
> Is it possible to truncate on the first line? I couldn't find another way of doing that.

It is not at the moment

> Is it something you consider adding in the future?

Possibly! I guess we want to find the sweet spot between giving consumers of the components flexibility, while also ensuring a certain level of "art direction" on how the component behaves.

@WordPress/gutenberg-design do you have any guidelines on text truncation for both label and help text in the dropdown menu?

Reply:
It looks like the `DropdownMenuItemLabel` gets truncated at one line even when `numberOfLines` is undefined, so at least as far as this PR is concerened, it seems like truncation is only an issue with the `DropdownMenuItemHelpText`:

<img width="629" alt="truncation" src="https://github.com/user-attachments/assets/ff2e3bd4-51b2-45bd-9354-2a2c4e2972b9">

Reply:
The main issue I see with truncating to two lines in this use case is that it creates items with different widths depending on the size of the value, which feels a bit unusual.

Reply:
I was trying to remove the `numberOfLines` property and it seems there is a problem when I have a long string without spaces (it could be a URL, for example).

**With `numberOfLines` set to 1** 

![Screenshot 2024-07-25 at 11 57 56](https://github.com/user-attachments/assets/8015d49b-4ada-407f-89ed-60b91583394e)

**Without `numberOfLines`**

![Screenshot 2024-07-25 at 11 55 35](https://github.com/user-attachments/assets/e4d1b6e9-4f59-4819-808e-6a289a44d477)

---
Would be okay to keep it and figure out how to solve the problem in a follow-up pull request?

Reply:
We spotted this during testing yesterday, and @ciampo filed #63916, which should solve that exact problem.

Reply:
It looks like it got solved as you typed. That was a quick one! :)

Reply:
Yep, it was just merged - feel free to rebase and try it out!

Reply:
Great! I just rebased and the problem is not there anymore. Thank you! 🙂 I removed the `numberOfLines` prop until we figure out if it should be accepted as a parameter.

Reply:
Sorry, was AFK over the past days — the truncation for the label should be solved.

Apart from that, the item label and the item help text components intentionally truncate text to 1 line (for the label) and 2 lines (for the help text). It's a design decision, as we didn't want the content of a menu item to become very long — user should try to have short labels and help text.

If we think that such truncation is too limiting, please leave feedback in a separate issue and we'll discuss it there!